src/transport.rs

```
@@ -0,0 +1,423 @@
+    /// timeout duration, it should return [`WorkerShutdown::Success`].
+    fn shutdown(&self, timeout: std::time::Duration) -> TransportShutdown;
+
+    /// Constructs an HTTP request for the provided [`sys::Envelope`] with the
+    /// DSN that was registered with the SDK.
+    ///
+    /// The return value has all of the necessary pieces of data to create a
+    /// HTTP request with the HTTP client of your choice:
+    ///
+    /// * The uri to send the request to
+    /// * The headers that must be set
+    /// * The body of the request
+    ///
+    /// The `content-length` header is already set for you, though some HTTP
+    /// clients will automatically set it for you in some cases, which should
+    /// be fine.
+    ///
+    /// The `body` in the request is a [`Envelope`], which implements `AsRef<[u8]`
+    /// to retrieve the actual bytes that should be sent as the body
+    ///
+    /// # Errors
+    /// Can fail if the envelope can't be serialized, or there is an invalid
+    /// header value
+    fn convert_to_request(
+        dsn: &Dsn,
+        envelope: PostedEnvelope,
+    ) -> Result<SentryRequest, &'static str>
+    where
+        Self: Sized,
+    {
+        let mut builder = http::Request::builder();
+
+        {
+            let headers = builder
+                .headers_mut()
+                .ok_or_else(|| "unable to mutate headers")?;
+            headers.insert(
+                "user-agent",
+                USER_AGENT
+                    .parse()
+                    .map_err(|_| "failed to parse user agent")?,
+            );
+            headers.insert(
+                "content-type",
+                ENVELOPE_MIME
+                    .parse()
+                    .map_err(|_| "failed to parse MIME type")?,
+            );
+            headers.insert(
+                "accept",
+                "*/*".parse().map_err(|_| "failed to parse accept")?,
+            );
+        }
```

As far as I can see all of this should be unable to fail.
In the rest of the crate I handled most of the failures that could stem from `sentry-native` as panics. For example invalid UTF-8 or invalid MessagePack.

Any particular reason we shouldn't let it panic here too? What do you think about this issue in general?

---

Just to clarify, I also mean the other errors that you build below, specifically the code that I marked here is irrelevant, sorry about that.

---

This was definitely overkill, because yes, none of these should fail. I'll just do unwraps.

---

I'm looking through the C implementation of all the other failures you took into consideration, will remove some of them when I commit.